daemons/ipa-sam/ipa_sam.c

```diff
@@ -1974,43 +1980,82 @@ static bool handle_cross_realm_princs(struct ipasam_private *ipasam_state,
 
 	trusted_dn = trusted_domain_dn(tmp_ctx, ipasam_state, domain);
 
-	princ_l = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s", remote_realm,
-			ipasam_state->realm);
-	princ_r = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
-			ipasam_state->realm, remote_realm);
+	princ_l = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
+				  remote_realm, ipasam_state->realm);
+	princ_l_tdo = talloc_asprintf(tmp_ctx, "%s$@%s",
+				      flat_name, ipasam_state->realm);
+	l_tdo_alias = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
+				      flat_name, ipasam_state->realm);
 
-	princ_tdo = talloc_asprintf(tmp_ctx, "%s$@%s",
-			ipasam_state->flat_name, remote_realm);
+	princ_r = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
+				  ipasam_state->realm, remote_realm);
+	princ_r_tdo = talloc_asprintf(tmp_ctx, "%s$@%s",
+				      ipasam_state->flat_name, remote_realm);
 
-	saltprinc_tdo = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
+	r_tdo_alias = talloc_asprintf(tmp_ctx, "krbtgt/%s@%s",
 			ipasam_state->flat_name, remote_realm);
 
-	if (trusted_dn == NULL || princ_l == NULL ||
-	    princ_r == NULL || princ_tdo == NULL || saltprinc_tdo == NULL) {
+	if (trusted_dn == NULL || princ_l == NULL || princ_l_tdo == NULL ||
+		l_tdo_alias == NULL || princ_r == NULL || princ_r_tdo == NULL ||
+		r_tdo_alias == NULL) {
 		ok = false;
 		goto done;
 	}
 
 	switch (mod) {
 		case SET_PRINC:
-			/* Create Kerberos principal for inbound trust, enabled by default */
-			ok   = set_krb_princ(ipasam_state, tmp_ctx, princ_r, NULL, pwd, trusted_dn, KRB_PRINC_CREATE_DEFAULT);
-			/* Create Kerberos principal corresponding to TDO in AD for SSSD usage, disabled by default */
-			ok |= set_krb_princ(ipasam_state, tmp_ctx, princ_tdo, saltprinc_tdo, pwd, trusted_dn,
-					    KRB_PRINC_CREATE_DISABLED | KRB_PRINC_CREATE_AGENT_PERMISSION);
+			/* We must use two sets by two principals here because
+			 * they are used for different needs and must have
+			 * different salts */
+
+			/* INBOUND TRUST */
+			if ((trust_direction & LSA_TRUST_DIRECTION_INBOUND) != 0) {
+				/* First: krbtgt/<OUR REALM>@<REMOTE REALM>, enabled by default
+				 * in case of the inboud trust */
+				ok   = set_krb_princ(ipasam_state, tmp_ctx, princ_r, NULL,
+						     pwd_outgoing, trusted_dn, KRB_PRINC_CREATE_DEFAULT);
+
+				/* Second: <OUR FLATNAME$>@<REMOTE REALM> is only used
+				 * for SSSD to be able to talk to AD DCs but it has to
+				 * have canonical name set to <OUR FLATNAME>$ because
+				 * this is the salt used by AD DCs when using this
+				 * principal, otherwise authentication will fail.
+				 *
+				 * *disable* use of this principal on our side as it is
+				 * only used to retrieve trusted domain credentials by
+				 * AD Trust Agents across the IPA topology */
+				ok |= set_krb_princ(ipasam_state, tmp_ctx, r_tdo_alias, princ_r_tdo,
```

Comment:
This trick doesn't work. bool false is ``0`` and bool true is ``1``. The boolean or operator ``|`` won't reset ``ok`` to false when one of the ``set_krb_princ`` calls fail.